src/version_check.py

```diff
@@ -0,0 +1,244 @@
+    if charm_path.split(":")[0] == LOCAL_BUILT_CHARM_PREFIX:
+        logger.debug("Charm is locally built. Cannot determine revision number.")
+        return 0
+
+    # charm_path is of the format ch:amd64/jammy/<charm-name>-<revision number>
+    revision = charm_path.split("-")[-1]
+    return int(revision)
+
+
+class CrossAppVersionChecker(Object):
+    """Verifies versions across multiple integrated applications."""
+
+    def __init__(
+        self,
+        charm,
+        version: int,
+        relations_to_check: List[str],
+        version_validity_range: Optional[Dict] = None,
+    ) -> None:
+        """Constructor for CrossAppVersionChecker object.
+
+        Args:
+            charm: charm to inherit from
+            version: (int), the current version of the desired attribute of the charm
+            relations_to_check: (List), a list of relations who should have compatible versions
+                with the current charm
+            version_validity_range: (Optional Dict), a list of ranges for valid version ranges.
+                If not provided it is assumed that relations on the provided interface must have
+                the same version.
+        """
+        self.charm = charm
+        super().__init__(self.charm, None)
+        # Future PR: upgrade this to a dictionary name versions
+        self.version = version
+        self.relations_to_check = relations_to_check
+
+        for rel in relations_to_check:
+            self.framework.observe(
+                charm.on[rel].relation_created,
+                self.set_version_on_relation_created,
+            )
+
+        # this feature has yet to be implemented, MongoDB does not need it and it is unclear if
+        # this will be extended to other charms. If this code is extended to other charms and
+        # there is a valid usecase we will use the `version_validity_range` variable in the
+        # function `get_invalid_versions`
+        self.version_validity_range = version_validity_range
+
+    def get_invalid_versions(self) -> List[Tuple[str, int]]:
+        """Returns a tuple of (app name, version number), if the version number mismatches.
+
+        Mismatches are decided based on version_validity_range, if version_validity_range is not
+        provided, then the mismatch is expected to match this current app's version number.
+
+        Raises:
+            NoVersionError.
+        """
+        try:
+            invalid_relations = []
+            for relation_name in self.relations_to_check:
+                for relation in self.charm.model.relations[relation_name]:
+                    related_version = relation.data[relation.app][VERSION_CONST]
+                    if int(related_version) != self.version:
+                        invalid_relations.append((relation.app.name, related_version))
+        except KeyError:
+            raise NoVersionError(f"Expected {relation.app.name} to have version info.")
```

Comment:
My reasoning is that relation X isn't necessarily invalid, it could be valid. But we don't know since we don't have any of its version info. I believe this is a separate case.

This can happen in the following scenarios:
1. Provider uses `self.version_checker = CrossAppVersionChecker( ... )` in src/charm.py, but Requirer does NOT. (Developer mistake)
2. The integration on is made to a much older version of a charm, in which the newer version *does* share the version info